sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1860,7 +1860,12 @@ class Dataset[T] private[sql](
    * @since 1.6.0
    */
   @Experimental
-  def filter(func: T => Boolean): Dataset[T] = mapPartitions(_.filter(func))
+  def filter(func: T => Boolean): Dataset[T] = {
+    val deserializer = unresolvedTEncoder.deserializer
+    val function = Literal.create(func, ObjectType(classOf[Function1[T, Boolean]]))
+    val condition = Invoke(function, "apply", BooleanType, deserializer :: Nil)
+    withTypedPlan(Filter(condition, logicalPlan))
```

Comment:
Previously we need to deserialize input row, call function and serialize to row, but now we don't need to do the final serialization, it should be faster.